function-images/tests/tuning-halving/driver/main.py

```diff
@@ -0,0 +1,264 @@
+	n_samples = 1000
+	n_features = 1024 
+	X, y = datasets.make_classification(n_samples,
+	                                    n_features,
+	                                    n_redundant=0,
+	                                    n_clusters_per_class=2,
+	                                    weights=[0.9, 0.1],
+	                                    flip_y=0.1,
+	                                    random_state=42)
+	return {'features': X, 'labels': y}
+
+def generate_hyperparam_sets(param_config):
+	keys = list(param_config.keys())
+	values = [param_config[k] for k in keys]
+
+	for elements in itertools.product(*values):
+		yield dict(zip(keys, elements))
+
+class GreeterServicer(helloworld_pb2_grpc.GreeterServicer):
+    def __init__(self, transferType, XDTconfig=None):
+
+        self.benchName = 'vhive-tuning'
+        self.transferType = transferType
+        if transferType == S3:
+            self.s3_client = boto3.resource(
+                service_name='s3',
+                region_name=os.getenv("AWS_REGION", 'us-west-1'),
+                aws_access_key_id=AWS_ID,
+                aws_secret_access_key=AWS_SECRET
+            )
+        elif transferType == XDT:
+            if XDTconfig is None:
+                log.fatal("Empty XDT config")
+            self.XDTconfig = XDTconfig
+
+    def put(self, obj, key):
+        msg = "Driver uploading object with key '" + key + "' to " + self.transferType
+        log.info(msg)
+        with tracing.Span(msg):
+            pickled = pickle.dumps(obj)
+            if self.transferType == S3:
+                s3object = self.s3_client.Object(bucket_name=self.benchName, key=key)
+                s3object.put(Body=pickled)
+            elif self.transferType == XDT:
+                log.fatal("XDT is not supported")
+
+        return key
+
+    def get(self, key):
+        msg = "Driver gets key '" + key + "' from " + self.transferType
+        log.info(msg)
+        with tracing.Span(msg):
+            response = None
+            if self.transferType == S3:
+                obj = self.s3_client.Object(bucket_name=self.benchName, key=key)
+                response = obj.get()
+            elif self.transferType == XDT:
+                log.fatal("XDT is not yet supported")
+
+        return pickle.loads(response['Body'].read())
+
+    def handler_broker(self, event, context):
+        dataset = generate_dataset()
+        hyperparam_config = {
+            'model': 'RandomForestRegressor',
+            'params': {
+                'n_estimators': [5, 10, 20],
+                'min_samples_split': [2, 4],
+                'random_state': [42]
+            }
+        }
+        models_config = {
+            'models': [
+                {
+                    'model': 'RandomForestRegressor',
+                    'params': hyperparam
+                } for hyperparam in generate_hyperparam_sets(hyperparam_config['params'])
+            ]
+        }
+        self.put(dataset, 'dataset_key')
```

Comment:
BUG: need to record the key that `put()` returns and pass it. make sure that every function does it. otherwise XDT would not work